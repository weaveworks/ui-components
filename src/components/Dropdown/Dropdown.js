import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import { map, find, flattenDepth, isArray } from 'lodash';

const WIDTH = '256px';
const HEIGHT = '36px';

const Item = styled.div`
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding: 0 12px;
  cursor: pointer;
`;

const Popover = styled.div`
  position: absolute;
  background-color: ${props => props.theme.colors.white};
  border: 1px solid ${props => props.theme.colors.gray200};
  border-radius: ${props => props.theme.borderRadius.soft};
  z-index: ${props => props.theme.layers.dropdown};
  box-shadow: ${props => props.theme.boxShadow.light};
  margin-top: 4px;
  width: ${props => props.width}px;
  box-sizing: border-box;
  padding: 6px 0;
`;

const Overlay = styled.div`
  z-index: ${props => props.theme.layers.dropdown};
  position: fixed;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
`;

const ItemWrapper = Item.extend`
  line-height: ${HEIGHT};
  color: ${props => (props.selected ? props.theme.colors.blue400 : props.theme.textColor)};
  min-height: ${HEIGHT};

  &:hover {
    background-color: ${props => props.theme.colors.gray50};
  }
`;

const Divider = styled.div`
  margin: 6px 0;
  border-bottom: 1px solid ${props => props.theme.colors.gray200};
`;

const SelectedItem = Item.extend`
  border-radius: ${props => props.theme.borderRadius.soft};
  background-color: ${props => props.theme.colors.white};
  border: 1px solid ${props => props.theme.colors.gray600};
  display: flex;

  ${Item} {
    padding: 0;
  }

  div:last-child {
    margin-left: auto;
  }
`;

const SelectedItemIcon = styled.span`
  padding-left: 1em;
  float: right;
  line-height: ${HEIGHT} !important;
`;

const StyledDropdown = component => styled(component)`
  height: ${HEIGHT};
  line-height: 34px;
  position: relative;
  width: ${WIDTH};
  box-sizing: border-box;
`;

const DefaultToggleView = ({ onClick, selectedLabel }) => (
  <SelectedItem onClick={onClick} className="dropdown-toggle">
    <Item>{selectedLabel}</Item>
    <div>
      <SelectedItemIcon className="fa fa-caret-down" />
    </div>
  </SelectedItem>
);

/**
 * A selectable drop-down menu.
 * ```javascript
 *const items = [
 *  {
 *    value: 'first-thing',
 *    label: 'First Thing',
 *  },
 *  {
 *    value: 'second-thing',
 *    label: 'Second Thing',
 *  },
 * ];
 *
 * <Dropdown items={items} />
 * ```
 *
 * You  may also add `null` for dividers or provide groups that will be separated
 * ```javascript
 *const items = [
 *  [
 *    {
 *      value: 'first-thing',
 *      label: 'First Thing',
 *    },
 *    {
 *      value: 'second-thing',
 *      label: 'Second Thing',
 *    },
 *  ],
 *  [
 *    {
 *      value: 'another-thing',
 *      label: 'Another Thing',
 *    },
 *  ]
 * ];
 * ```
 */
class Dropdown extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      isOpen: false,
    };

    this.element = React.createRef();

    this.handleChange = this.handleChange.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.handleBgClick = this.handleBgClick.bind(this);
  }

  handleChange(ev, value, label) {
    this.setState({ isOpen: false });
    if (this.props.onChange) {
      this.props.onChange(ev, value, label);
    }
  }

  handleClick() {
    this.setState({ isOpen: true });
  }

  handleBgClick() {
    this.setState({ isOpen: false });
  }

  divide(items) {
    if (!items || !isArray(items[0])) {
      return items;
    }

    return flattenDepth(items.map(it => [null, it]), 2).slice(1);
  }

  render() {
    const { items, value, className, placeholder } = this.props;
    const { isOpen } = this.state;
    const divided = this.divide(items);
    // If nothing is selected, use the placeholder, else use the first item.
    const currentItem =
      find(divided, i => i && i.value === value) ||
      (placeholder ? { label: placeholder, value: null } : divided && divided[0]);
    const label = currentItem && currentItem.label;
    const Component = this.props.withComponent;

    return (
      <div className={className} title={label} ref={this.element}>
        <Component selectedLabel={label} onClick={this.handleClick} />
        {isOpen && (
          <div>
            <Overlay onClick={this.handleBgClick} />
            <Popover width={this.element.current.offsetWidth}>
              {map(
                divided,
                (item, index) =>
                  item ? (
                    <ItemWrapper
                      className="dropdown-item"
                      key={item.value}
                      onClick={ev => this.handleChange(ev, item.value, item.label)}
                      selected={item.value === value}
                      title={item && item.label}
                    >
                      {item.label}
                    </ItemWrapper>
                  ) : (
                    <Divider key={index} />
                  )
              )}
            </Popover>
          </div>
        )}
      </div>
    );
  }
}

const itemPropType = PropTypes.shape({
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  label: PropTypes.string,
});

Dropdown.propTypes = {
  /**
   * Array of items (or groups of items) that will be selectable.
   * `value` should be an internal value,
   * `label` is what will be displayed to the user.
   */
  items: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.arrayOf(itemPropType), itemPropType]))
    .isRequired,
  /**
   * The value of the currently selected item. This much match a value in the `items` prop.
   * If no value is provided, the first elements's value will be used.
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * A handler function that will run when a value is selected.
   */
  onChange: PropTypes.func,
  /**
   * The initial text that will be display before a user selects an item.
   */
  placeholder: PropTypes.string,

  /**
   * A custom component to replace the default toggle view.
   * The properties `selectedLabel` and `onClick` are provided. `onClick` needs to be incorporated
   * to make the dropdown list toggle.
   */
  withComponent: PropTypes.func,
};

Dropdown.defaultProps = {
  withComponent: DefaultToggleView,
};

export default StyledDropdown(Dropdown);
