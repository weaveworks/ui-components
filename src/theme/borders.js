import { kebabCase, forEach } from 'lodash';

export const borderRadius = {
  circle: '50%',
  none: '0',
  soft: '2px',
};

// Collects all theme border radii as SCSS vars.
export function themeBorderRadiiAsScss() {
  const themeBorderRadii = [];

  forEach(borderRadius, (value, name) => {
    themeBorderRadii.push(`$border-radius-${kebabCase(name)}: ${value}`);
  });

  return themeBorderRadii;
}
