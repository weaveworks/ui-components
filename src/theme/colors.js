import { isString, kebabCase, forEach } from 'lodash';

export const colors = {
  // Neutral Colors
  black: 'hsl(0, 0%, 10%)', // #1a1a1a
  blue50: 'hsl(191, 100%, 97%)', // #f0fcff
  blue200: 'hsl(191, 100%, 80%)', // #99ecff
  blue400: 'hsl(191, 100%, 50%)', // #00d2ff
  blue600: 'hsl(191, 100%, 40%)', // #00a7cc
  blue700: 'hsl(191, 100%, 35%)', // #0092b3
  blue800: 'hsl(191, 100%, 30%)', // #007d99
  graphThemes: {
    blue: [
      '#c7e9b4',
      '#7ecdbb',
      '#1eb5eb',
      '#1d91bf',
      '#235fa9',
      '#253393',
      '#084181',
    ],
    mixed: [
      '#c7e9b4',
      '#c1d4e7',
      '#7ecdbb',
      '#9fbddb',
      '#1eb5eb',
      '#8d95c6',
      '#1d91bf',
      '#8282ab',
      '#235fa9',
      '#89429e',
      '#253393',
      '#800f7a',
      '#084181',
      '#0b0533',
    ],
    purple: [
      '#c1d4e7',
      '#9fbddb',
      '#8d95c6',
      '#8282ab',
      '#89429e',
      '#800f7a',
      '#0b0533',
    ],
  }, // Used by PrometheusGraph component
  gray50: 'hsl(0, 0%, 96%)', // #f4f4f4
  gray100: 'hsl(0, 0%, 90%)', // #e6e6e6
  gray200: 'hsl(0, 0%, 80%)', // #cccccc

  // #737373
  gray600: 'hsl(0, 0%, 45%)',
  green500: 'hsl(161, 54%, 48%)', // #38bd93
  orange500: 'hsl(13, 100%, 50%)', // #ff3700
  orange600: 'hsl(13, 100%, 40%)', // #cc2c00
  orange700: 'hsl(13, 100%, 35%)', // #b32700
  // Accent Colors
  // #992100
  orange800: 'hsl(13, 100%, 30%)',
  promQL: {
    attrName: '#00a4db',
    /**
     * GHColors theme by Avi Aryan (http://aviaryan.in)
     * Inspired by Github syntax coloring
     */
    comment: '#bbbbbb',
    deleted: '#9a050f',
    entity: '#36acaa',
    function: '#dc322f',
    metricName: '#2aa198',
    punctuation: '#393a34',
    // Dropdown colors,
    salmon: '#ff7c7c',
    string: '#e3116c',
    tag: '#00009f',
  }, // PromQL
  purple25: 'hsl(240, 20%, 98%)', // #fafafc
  purple50: 'hsl(240, 20%, 95%)', // #eeeef4
  purple100: 'hsl(240, 20%, 90%)', // #dfdfea
  purple200: 'hsl(240, 20%, 75%)', // #b1b1cb
  purple300: 'hsl(240, 20%, 65%)', // #9494b8
  purple400: 'hsl(240, 20%, 60%)', // #8585ad

  purple500: 'hsl(240, 20%, 50%)', // #666699
  purple600: 'hsl(240, 20%, 45%)', // #5b5b88
  purple700: 'hsl(240, 20%, 35%)', // #47476b
  purple800: 'hsl(240, 20%, 30%)', // #3d3d5c
  // Primary Colors
  // #32324b
  purple900: 'hsl(240, 20%, 25%)',

  // Third-party specific colors - not to be used in the theme!
  thirdParty: {
    azure: '#3769bb',
    // Google single-click login
    cornflowerBlue: '#4285f4',
  },
  // #ffffff
  white: 'hsl(0, 0%, 100%)',
  // #d4ab27
  yellow500: 'hsl(46, 69%, 49%)',
};

// Flattens and collects all theme colors as SCSS vars
export function themeColorsAsScss() {
  const ignoreKeys = ['graphThemes'];
  const themeColors = [];

  forEach(colors, (value, name) => {
    const colorPrefix = `$color-${kebabCase(name)}`;
    if (ignoreKeys.includes(name)) return;

    if (isString(value)) {
      themeColors.push(`${colorPrefix}: ${value}`);
    } else {
      forEach(value, (innerValue, subname) => {
        if (ignoreKeys.includes(subname)) return;
        themeColors.push(`${colorPrefix}-${kebabCase(subname)}: ${innerValue}`);
      });
    }
  });

  return themeColors;
}
