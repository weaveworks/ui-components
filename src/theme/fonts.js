import { kebabCase, forEach } from 'lodash';

export const fontFamilies = {
  monospace: "'Roboto Mono', monospace",
  regular: "'proxima-nova', Helvetica, Arial, sans-serif",
};

export const fontSizes = {
  extraLarge: '32px',
  huge: '48px',
  large: '22px',
  normal: '16px',
  small: '14px',
  tiny: '12px',
};

// Collects all theme font vars as SCSS vars.
export function themeFontsAsScss() {
  const themeFonts = [];

  forEach(fontFamilies, (value, name) => {
    themeFonts.push(`$font-family-${kebabCase(name)}: ${value}`);
  });

  forEach(fontSizes, (value, name) => {
    themeFonts.push(`$font-size-${kebabCase(name)}: ${value}`);
  });

  return themeFonts;
}
