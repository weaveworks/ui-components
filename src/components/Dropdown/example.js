import React from 'react';

import { Example, Info } from '../../utils/example';
import { Grid, GridRow as Row, GridColumn as Column } from '../Grid';
import Button from '../Button';

import Dropdown from '.';

const generateItems = (hash = '') => [
  {
    label: 'First Thing',
    value: `first-thing${hash}`,
  },
  {
    label: 'Second Thing',
    value: `second-thing${hash}`,
  },
  {
    label: 'Super long thing, this should get truncated',
    value: `third-thing${hash}`,
  },
];

const items = generateItems();

const groups = [
  [
    {
      label: 'A: First Thing',
      value: 'a-first',
    },
    {
      label: 'A: Second Thing',
      value: 'a-second',
    },
  ],
  [
    {
      label: 'B: First Thing',
      value: 'b-first',
    },
    {
      label: 'B: Second Thing',
      value: 'b-second',
    },
  ],
];

const divided = [
  {
    label: 'C: First Thing',
    value: 'c-first',
  },
  null,
  {
    label: 'D: First Thing',
    value: 'd-first',
  },
];

const DropdownButton = ({ onClick }) => (
  <Button onClick={onClick}>
    More actions <i className="fa fa-caret-down" />
  </Button>
);

export default class DropdownExample extends React.Component {
  constructor() {
    super();
    this.state = {
      selected: null,
    };
  }

  handleChange = (ev, value) => {
    this.setState(() => ({ selected: value }));
  };

  render() {
    return (
      <Grid>
        <Row>
          <Column span={6}>
            <Example>
              <Info>Default dropdown</Info>
              <Dropdown
                items={[
                  ...generateItems('a'),
                  ...generateItems('b'),
                  ...generateItems('c'),
                  ...generateItems('d'),
                ]}
                value={this.state.selected}
                onChange={this.handleChange}
              />
            </Example>
          </Column>
          <Column span={6}>
            <Example>
              <Info>Grouped items (passed as subarrays)</Info>
              <Dropdown
                items={groups}
                value={this.state.selected}
                onChange={this.handleChange}
              />
            </Example>
          </Column>
        </Row>
        <Row>
          <Column span={6}>
            <Example>
              <Info>Grouped items (with null as dividers)</Info>
              <Dropdown
                items={divided}
                value={this.state.selected}
                onChange={this.handleChange}
              />
            </Example>
          </Column>
          <Column span={6}>
            <Example>
              <Info>With placeholder</Info>
              <Dropdown
                items={items}
                placeholder="Select an item"
                value={this.state.selected}
                onChange={this.handleChange}
              />
            </Example>
          </Column>
        </Row>
        <Row>
          <Column span={6}>
            <Example>
              <Info>
                With <code>Button</code> component
              </Info>
              <Dropdown
                withComponent={DropdownButton}
                items={divided}
                value={this.state.selected}
                onChange={this.handleChange}
              />
            </Example>
          </Column>
        </Row>
        <div style={{ height: '70px' }} />
      </Grid>
    );
  }
}
