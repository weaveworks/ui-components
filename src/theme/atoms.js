import { colors } from './colors';

// component-specific
export const atoms = {
  Alert: {
    error: {
      background: colors.orange600,
      color: colors.white,
    },
    info: {
      background: colors.blue600,
      color: colors.white,
    },
    success: {
      background: colors.green500,
      color: colors.white,
    },
    warning: {
      background: colors.yellow500,
      color: colors.white,
    },
  },
  Button: {
    danger: {
      background: colors.orange600,
      color: colors.white,
      hoverBackground: colors.orange700,
      hoverColor: colors.white,
    },
    default: {
      background: colors.white,
      color: colors.black,
      hoverBackground: colors.gray50,
      hoverColor: colors.purple800,
    },
    disabled: {
      background: colors.gray50,
      color: colors.gray600,
      hoverBackground: colors.gray50,
      hoverColor: colors.gray600,
    },
    primary: {
      background: colors.blue700,
      color: colors.white,
      hoverBackground: colors.blue800,
      hoverColor: colors.white,
    },
  },
};
