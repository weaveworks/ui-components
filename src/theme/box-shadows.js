import { colors } from './colors';

export const boxShadow = {
  heavy: '0 10px 20px rgba(0,0,0,0.19), 0 6px 6px rgba(0,0,0,0.23)',
  light: '0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24)',
  none: 'none',
  selected: `0px 0px 2px 2px ${colors.purple400}`,
};
