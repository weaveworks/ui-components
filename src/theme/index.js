import { colors, themeColorsAsScss } from './colors';
import { fontSizes, fontFamilies, themeFontsAsScss } from './fonts';
import { borderRadius, themeBorderRadiiAsScss } from './borders';
import { layers, themeLayersAsScss } from './layers';
import { spacing, themeSpacingsAsScss } from './spacings';
import { overlayIconSize, textColor, themeMiscVarsAsScss } from './misc';
import { boxShadow } from './box-shadows';
import { atoms } from './atoms';

export function themeVarsAsScss() {
  const themeVariables = []
    .concat(themeColorsAsScss())
    .concat(themeLayersAsScss())
    .concat(themeFontsAsScss())
    .concat(themeBorderRadiiAsScss())
    .concat(themeSpacingsAsScss())
    .concat(themeMiscVarsAsScss());
  return `${themeVariables.join('; ')};`;
}

const theme = {
  // Component-specific
  atoms,

  // Borders
  borderRadius,

  // Box shadows
  boxShadow,

  // Colors
  colors,

  // Fonts
  fontFamilies,
  fontSizes,

  // z-index layers
  layers,

  // Misc
  overlayIconSize,

  // Spacings
  spacing,
  textColor,
};

export default theme;
