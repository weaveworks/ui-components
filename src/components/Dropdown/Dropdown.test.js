import React from 'react';
import renderer from 'react-test-renderer';
import styled from 'styled-components';

import 'jest-styled-components';
import { withTheme } from '../../utils/theme';

import Dropdown from './Dropdown';

describe('<Dropdown />', () => {
  const items = [
    {
      label: 'First Thing',
      value: 'first-thing',
    },
    {
      label: 'Second Thing',
      value: 'second-thing',
    },
    {
      label: 'Super long thing, this should get truncated',
      value: 'third-thing',
    },
  ];

  describe('snapshots', () => {
    it('renders default', () => {
      const tree = renderer
        .create(withTheme(<Dropdown items={items} />))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders a placeholder', () => {
      const tree = renderer
        .create(
          withTheme(<Dropdown placeholder="Select an item" items={items} />)
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it('is extendable', () => {
    // Make sure a styled-component gets exported
    expect(typeof styled(Dropdown)).toEqual('function');
    expect(Dropdown.displayName).toEqual('Dropdown');
  });
});
