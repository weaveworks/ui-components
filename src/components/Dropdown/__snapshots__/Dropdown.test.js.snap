// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> snapshots renders a placeholder 1`] = `
.c3 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding-right: 28px;
  padding-left: 12px;
  cursor: pointer;
}

.c1 {
  height: 36px;
  box-sizing: border-box;
  border-radius: 2px;
  background-color: hsl(0,0%,100%);
  border: 1px solid hsl(0,0%,45%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 .c2 {
  padding: 0;
}

.c1 div:last-child {
  margin-left: auto;
}

.c4 {
  position: absolute;
  right: 10px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c0 {
  height: 36px;
  line-height: 34px;
  position: relative;
  width: 256px;
  box-sizing: border-box;
}

<div
  className="c0"
  title="Select an item"
>
  <div
    className="dropdown-toggle c1 c2 c3"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="c2 c3"
      disabled={false}
    >
      Select an item
    </div>
    <span
      className="fa fa-caret-down c4"
    />
  </div>
</div>
`;

exports[`<Dropdown /> snapshots renders default 1`] = `
.c3 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding-right: 28px;
  padding-left: 12px;
  cursor: pointer;
}

.c1 {
  height: 36px;
  box-sizing: border-box;
  border-radius: 2px;
  background-color: hsl(0,0%,100%);
  border: 1px solid hsl(0,0%,45%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 .c2 {
  padding: 0;
}

.c1 div:last-child {
  margin-left: auto;
}

.c4 {
  position: absolute;
  right: 10px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c0 {
  height: 36px;
  line-height: 34px;
  position: relative;
  width: 256px;
  box-sizing: border-box;
}

<div
  className="c0"
  title="First Thing"
>
  <div
    className="dropdown-toggle c1 c2 c3"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="c2 c3"
      disabled={false}
    >
      First Thing
    </div>
    <span
      className="fa fa-caret-down c4"
    />
  </div>
</div>
`;
